File: sdk-js/src/sender.ts
type QueueItem = { path: string; payload: unknown }

export class BackgroundSender {
  private readonly apiUrl: string
  private readonly headers: Record<string, string>
  private readonly queue: QueueItem[] = []
  private flushing = false

  constructor(apiUrl: string, token: string) {
    this.apiUrl = apiUrl.replace(/\/$/, "")
    this.headers = {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    }
  }

  enqueue(path: string, payload: unknown): void {
    if (this.queue.length > 1000) return // drop rather than grow unbounded
    this.queue.push({ path, payload })
    if (!this.flushing) this.drain()
  }

  async flush(): Promise<void> {
    while (this.queue.length > 0) {
      await this.drain()
    }
  }

  private async drain(): Promise<void> {
    if (this.flushing || this.queue.length === 0) return
    this.flushing = true
    const items = this.queue.splice(0, this.queue.length)
    await Promise.all(
      items.map(({ path, payload }) =>
        fetch(`${this.apiUrl}${path}`, {
          method: "POST",
          headers: this.headers,
          body: JSON.stringify(payload),
        }).catch(() => null), // never throw — agent must not crash
      ),
    )
    this.flushing = false
    if (this.queue.length > 0) this.drain()
  }
}

File: sdk-python/midlantics_a2a/sender.py
"""Background HTTP sender — queues payloads and ships them in a daemon thread."""
from __future__ import annotations

import json
import queue
import threading
import time
from typing import Any

import httpx

_STOP = object()


class BackgroundSender:
    def __init__(self, api_url: str, token: str, max_queue: int = 1000) -> None:
        self._api_url = api_url.rstrip("/")
        self._headers = {"Authorization": f"Bearer {token}", "Content-Type": "application/json"}
        self._q: queue.Queue = queue.Queue(maxsize=max_queue)
        self._thread = threading.Thread(target=self._run, daemon=True, name="a2a-sender")
        self._thread.start()

    # ── Public ────────────────────────────────────────────────────────────────

    def enqueue(self, path: str, payload: dict[str, Any]) -> None:
        try:
            self._q.put_nowait((path, payload))
        except queue.Full:
            pass  # drop rather than block the agent

    def flush(self, timeout: float = 5.0) -> None:
        self._q.join()

    def shutdown(self, timeout: float = 5.0) -> None:
        self._q.put(_STOP)
        self._thread.join(timeout=timeout)

    # ── Internal ──────────────────────────────────────────────────────────────

    def _run(self) -> None:
        with httpx.Client(timeout=10.0) as client:
            while True:
                item = self._q.get()
                if item is _STOP:
                    self._q.task_done()
                    break
                path, payload = item
                try:
                    client.post(
                        f"{self._api_url}{path}",
                        content=json.dumps(payload, default=str),
                        headers=self._headers,
                    )
                except Exception:
                    pass  # never crash the agent
                finally:
                    self._q.task_done()
